packages/core/src/diff/changes/directive-usage.ts

```diff
@@ -0,0 +1,874 @@
+        const { interfaceName, addedDirectiveName } = meta;
+        return createDirectiveChange(
+          ChangeType.DirectiveUsageInterfaceAdded,
+          `Directive '${addedDirectiveName}' was added to interface '${interfaceName}'`,
+          isOneOfDirective(addedDirectiveName)
+            ? CriticalityLevel.Dangerous
+            : CriticalityLevel.Breaking,
+          [interfaceName, addedDirectiveName].join('.'),
+        );
+      }
+      break;
+
+    case 'ENUM':
+      if ('enumName' in meta) {
+        const { enumName, addedDirectiveName } = meta;
+        return createDirectiveChange(
+          ChangeType.DirectiveUsageEnumAdded,
+          `Directive '${addedDirectiveName}' was added to enum '${enumName}'`,
+          CriticalityLevel.Dangerous,
+          [enumName, addedDirectiveName].join('.'),
+        );
+      }
+      break;
+
+    case 'ENUM_VALUE':
+      if ('enumName' in meta && 'enumValueName' in meta) {
+        const { enumName, enumValueName, addedDirectiveName } = meta;
+        return createDirectiveChange(
+          ChangeType.DirectiveUsageEnumValueAdded,
+          `Directive '${addedDirectiveName}' was added to enum value '${enumName}.${enumValueName}'`,
+          CriticalityLevel.Dangerous,
+          [enumName, enumValueName, addedDirectiveName].join('.'),
+        );
+      }
+      break;
+
+    case 'UNION':
+      if ('unionName' in meta && 'addedDirectiveName' in meta) {
+        const { unionName, addedDirectiveName } = meta;
+        return createDirectiveChange(
+          ChangeType.DirectiveUsageUnionMemberAdded,
+          `Directive '${addedDirectiveName}' was added to union member '${unionName}'`,
+          CriticalityLevel.Dangerous,
+          [unionName, addedDirectiveName].join('.'),
+        );
+      }
+      break;
+
+    case 'SCALAR':
+      if ('scalarName' in meta && 'addedDirectiveName' in meta) {
+        const { scalarName, addedDirectiveName } = meta;
+        return createDirectiveChange(
+          ChangeType.DirectiveUsageScalarAdded,
+          `Directive '${addedDirectiveName}' was added to scalar '${scalarName}'`,
+          isOneOfDirective(addedDirectiveName)
+            ? CriticalityLevel.Dangerous
+            : CriticalityLevel.Breaking,
+          [scalarName, addedDirectiveName].join('.'),
+        );
+      }
+      break;
+    case 'FIELD_DEFINITION':
+      if ('typeName' in meta && 'fieldName' in meta && 'addedDirectiveName' in meta) {
+        const { typeName, fieldName, addedDirectiveName } = meta;
+        return createDirectiveChange(
+          ChangeType.DirectiveUsageFieldDefinitionAdded,
+          `Directive '${addedDirectiveName}' was added to field '${typeName}.${fieldName}'`,
+          CriticalityLevel.Dangerous,
+          [typeName, fieldName, addedDirectiveName].join('.'),
+        );
+      }
+      break;
+  }
+
+  throw new Error(`Unhandled case: ${changeCase}`);
+}
+
+interface DirectiveUsageAddedController {
+  case:
```

Comment:
This could leverage TypeScript in a better way. 
You can build this as a map and then infer the expected type. 
Also, you can use `Kind` of GraphQL for that. 

Check this: https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgaWAOwCYBo4HEoCGYAFgIoAyAomgK4gAqAnmAKY75FnkCqawEaJq3aESFAPIAjAFYsAxjCFs4AQQDK9AHIQMLOAF84AMygQQcAOQBzUcQCOAGwsBuAFCuYzPakz0IKqCs6FjR4AF5EVzhouABtHwwAOkpNbgBZAH16AE0ABUoMgBFKADEASU0y+jLxTQBdAC48WwpqOiUsKJj49CTuStqsvILi8srq2sbmzgpefkEvTv13Ixo0BXm4YABncSMEgB4EuBYADxgQjG24AGsWRggjFF6-AKCQEJgcADUT88vrkgCNsYNpdAB+JrqLQ6PQAHzgtAcDgMAD4ABQ3XpNBI4ABuBAcNBYTW+AEomgSiXods9fP5AsFQj1MHVIjE4FAWDAaFA0HAqcTEsDQbDwYksZg4GEZbdem5lq5VusYJsMBA1GZucR0FYjr0-hdMNc7g8nglXoyPqFUXBMdi6dgBYTiTiXgz3p8WRg6mT2TEAPQBuAACRYXLgAHc9HICPyrNy4HIIBAUZ5WFs0EZwyE5HoutFgE90Ts9gl0QlkqlMjl8kVShUqjVNPiXSwyX6EAWOXAg6Gczh0NtgLo4DBiAR4AAiItTnDjvRTwUsKdwAhWz5ba66C4KFgYNfXDhiKi0BhebuGFgObY04ul-a9Cu9RL9ZtDOujRsTFvO6kd-0e17YMwy5Qc0GHUdx0nOAZyMOcx2IRdl1XddPVCLc4B3eQLgPYFphPOYBCUS9XEVPt6B1a5aSpEc1ylEBJzkJDtlcdVNQ+cddWfTAq3SD8RgbcZmxwNAWEjAiuDac9WHRJA0AID4minC4QQQ5dtiaBB9H0Ds3FcPsyieaMxygRgxwgOBGJgZiTjPOUpRgSzti1Li0CsE4b2UYB4CMAhgBvNiNVcnV3J4pIUn42tBLGJtalE8TJIkGQcKUOTEUUklYNUmAEKMYBryuLT9BwXSyX0oA